### PR TITLE
[Android] Implement using generic IM(invoke,read,write,subscribe) APIs

### LINKED ISSUE #25468: Use generic IM(invoke,read,write,subscribe) APIs in android chip-tool
Migrate light on/of, level cluster, sensor cluster, multiple admin cluster, operational credential cluster,basic cluster to use the generic IM API for cluster interaction.

### PR DESCRIPTION
**Pull request by joonhaengHeo**
Fix #25468 

The purpose of this PR is to:

1. Create Class that returns Cluster, Attribute, and Event ID (Using jinja)
2. Changed the function used in fragments of light on/of, level cluster, sensor cluster, multiple admin cluster, operational credential cluster, and basic cluster to Generic IM Function.

### CODE REVIEW

File: src/controller/java/templates/ChipIdLookup-java.zapt [hunk 1]
```diff
@@ -52,5 +52,98 @@ public final class ChipIdLookup {
     {{/all_user_clusters}}
     return "";
   }
+
+  {{#chip_client_clusters}}
+  public static final class {{asUpperCamelCase name}}Cluster {
```
Comment (andy31415) [inline src/controller/java/templates/ChipIdLookup-java.zapt:57]: This is a stand-alone class. Could you please move it into aseparate codegen so that we have individual generations?

Also please move it to jinja templates. I would prefer for java to standardize on jinja instead of zap.

**Comment by andy31415** [comment]
@joonhaengHeo please update the summary to explain the approach taken and how things are used. Ideally reviewers would read the summary to understand what to expect in the PR. `Fixes ...` forces me to go to a issue and even then, the issue describes just "use generic IM for ..." but does not explain how the implementation is done.

File: src/controller/java/templates/ChipIdLookup-java.zapt (hunk 1 above)
Comment (andy31415) [reply src/controller/java/templates/ChipIdLookup-java.zapt:57]: Ideally we would have separate java files for each cluster as well (jinja templates allow for this)

File: src/platform/android/java/chip/platform/NsdServiceFinderAndResolver.java [hunk 2]
```diff
@@ -132,7 +132,7 @@ public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
                 "Failed to resolve service '" + serviceInfo.getServiceName() + "': " + errorCode);
             chipMdnsCallback.handleServiceResolve(
                 serviceInfo.getServiceName(),
-                serviceInfo.getServiceType(),
+                NsdServiceFinderAndResolver.this.targetServiceInfo.getServiceType(),
```
Comment (andy31415) [inline src/platform/android/java/chip/platform/NsdServiceFinderAndResolver.java:135]: this seems like it should be a separate PR with separate description

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt [hunk 3]
```diff
@@ -44,15 +52,23 @@ class BasicClientFragment : Fragment() {
       childFragmentManager.findFragmentById(R.id.addressUpdateFragment) as AddressUpdateFragment
 
     binding.writeNodeLabelBtn.setOnClickListener { scope.launch {
-      sendWriteNodeLabelAttribute()
+      val tlvWriter = TlvWriter()
```
Comment (andy31415) [inline examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt:55]: This seems somewhat inconvenient to maintain. We probably would want poj-to-tlv and tlv-to-poj (but that is reasonably complex I would imagine).

Maybe add a TODO/issue for this.

**Comment by joonhaengHeo** [comment]
@andy31415 
Sorry for the confusion. The reason I posted this was not for the purpose of merging, but to share the current progress with others. As you said, I will write and upload it accurately at the time of PR.

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt (hunk 3 above)
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt:55]: Can you tell me what `poj-to-tlv` means?

File: src/controller/java/templates/ChipIdLookup-java.zapt (hunk 1 above)
Comment (joonhaengHeo) [reply src/controller/java/templates/ChipIdLookup-java.zapt:57]: Yes, I will review if it is possible with jinja.

File: src/platform/android/java/chip/platform/NsdServiceFinderAndResolver.java (hunk 2 above)
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdServiceFinderAndResolver.java:135]: I'm sorry. I wasn't crazy enough to check it out. This code was previously requested as #25645. It has not been merged yet, so it seems to have been added together by mistake.

File: src/controller/java/templates/ChipIdLookup-java.zapt (hunk 1 above)
Comment (joonhaengHeo) [reply src/controller/java/templates/ChipIdLookup-java.zapt]: @andy31415 
jinja file was created. Is there a separate script used for creation?
It seems that build errors occur in some cases after uploading the file.

Comment (joonhaengHeo) [reply src/controller/java/templates/ChipIdLookup-java.zapt]: @andy31415 
Do you know how to test after creating a jinja file?
If you have any related documents, it would be helpful if you could share them.

Comment (joonhaengHeo) [reply src/controller/java/templates/ChipIdLookup-java.zapt]: I don't know how to solve below error log.

`======================================================================
FAIL: test_generators (__main__.TestGenerators) (idl='inputs/several_clusters.matter', generator='java-class')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/connectedhomeip/connectedhomeip/out/coverage/../../scripts/py_matter_idl/matter_idl/test_generators.py", line 144, in run_test_cases
    generator.render(dry_run=False)
  File "/__w/connectedhomeip/connectedhomeip/scripts/py_matter_idl/matter_idl/generators/__init__.py", line 142, in render
    self.internal_render_all()
  File "/__w/connectedhomeip/connectedhomeip/scripts/py_matter_idl/matter_idl/generators/java/__init__.py", line 678, in internal_render_all
    self.internal_render_one_output(
  File "/__w/connectedhomeip/connectedhomeip/scripts/py_matter_idl/matter_idl/generators/__init__.py", line 181, in internal_render_one_output
    self.storage.write_new_data(output_file_name, rendered)
  File "/__w/connectedhomeip/connectedhomeip/out/coverage/../../scripts/py_matter_idl/matter_idl/test_generators.py", line 99, in write_new_data
    self.checker.assertEqual(
AssertionError: '/*\n[682 chars];\n\n\npublic class ClusterIDMapping {\n  publ[152689 chars]n}\n' != '/*\n[682 chars];\n\nimport java.util.HashMap;\nimport java.ut[6723 chars]}}\n' : Content of java/chip/devicecontroller/ClusterIDMapping.java

----------------------------------------------------------------------`

Comment (joonhaengHeo) [reply src/controller/java/templates/ChipIdLookup-java.zapt]: I found a way!

Comment (yunhanw-google) [reply src/controller/java/templates/ChipIdLookup-java.zapt]: @joonhaengHeo  have you figured out how to run and test jinja file?

Comment (joonhaengHeo) [reply src/controller/java/templates/ChipIdLookup-java.zapt]: I used "./scripts/codegen.py".
generator : java-class

e.g, : `./scripts/codegen.py --generator java-class ./scripts/py_matter_idl/matter_idl/tests/inputs/several_clusters.matter
`

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt (hunk 3 above)
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/BasicClientFragment.kt]: I added TODO comment.

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/SensorClientFragment.kt [hunk 4]
```diff
@@ -165,13 +165,15 @@ class SensorClientFragment : Fragment() {
       val deviceId = binding.deviceIdEd.text.toString().toULong().toLong()
       val endpointId = binding.endpointIdEd.text.toString().toInt()
       val clusterName = binding.clusterNameSpinner.selectedItem.toString()
-      val clusterSubscribe = CLUSTERS[clusterName]!!["subscribe"] as (Long, Int, ReadCallback) -> Unit
+      val clusterId = CLUSTERS[clusterName]!!["clusterId"] as Long
+      val attributeId = CLUSTERS[clusterName]!!["attributeId"] as Long
       val device = ChipClient.getConnectedDevicePointer(requireContext(), deviceId)
-      val callback = makeReadCallback(clusterName, true)
+      val callback = makeReadCallback(clusterName, false)
```
Comment (yunhanw-google) [inline examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/SensorClientFragment.kt:171]: readability: "xxx=" false, what does false reprents?

**Review by yunhanw-google — APPROVED** [review]

File: src/controller/java/BUILD.gn [hunk 5]
```diff
@@ -215,6 +215,7 @@ android_library("java") {
   data_deps = [ ":jni" ]
 
   sources = [
+    "generated/java/chip/devicecontroller/ClusterIDMapping.java",
     "generated/java/chip/devicecontroller/ClusterReadMapping.java",
```
Comment (yunhanw-google) [inline src/controller/java/BUILD.gn:219]: technically   android chip-tool don't depends on these old generated file, right? We just keep them here in case some folks still need them.

"generated/java/chip/devicecontroller/ClusterReadMapping.java",
"generated/java/chip/devicecontroller/ClusterWriteMapping.java",
"src/chip/clusterinfo/ClusterCommandCallback.java",

Comment (joonhaengHeo) [reply src/controller/java/BUILD.gn:219]: I know that the ".java" file is not currently being created in the out folder.
Do you know anything else?

image

File: examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/SensorClientFragment.kt (hunk 4 above)
Comment (joonhaengHeo) [reply examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/SensorClientFragment.kt]: Sorry. It seems that the part that was false when tested was uploaded as it is.
When true, the graph is displayed; when false, the graph is not displayed.

**Comment by bzbarsky-apple** [comment]
@joonhaengHeo Unfortunately, this is failing Androir CI (probably because of a bad merge against other changes?).  https://github.com/project-chip/connectedhomeip/pull/26704 created to revert this so we can get the tree back into a passing-CI state.
